drivers/mpsl/flash_sync/flash_sync_mpsl.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <errno.h>
+
+#include <mpsl.h>
+#include <mpsl_timeslot.h>
+
+#include "soc_flash_nrf.h"
+
+#define LOG_LEVEL CONFIG_SOC_FLASH_NRF_RADIO_SYNC_MPSL_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(flash_sync_mpsl);
+
+struct mpsl_context {
+	struct k_sem sem;  /* Synchronization semamphore. */
+	mpsl_timeslot_session_id_t session_id; /* Timeslot session ID. */
+	uint32_t duration; /* Requested duration. */
+	/* Argument passed to nrf_flash_sync_exe(). */
+	struct flash_op_desc *op_desc;
+	mpsl_timeslot_request_t timeslot_request;
+	/* Return parameter for the timeslot session. */
+	mpsl_timeslot_signal_return_param_t return_param;
+	int status; /* Return value for nrf_flash_sync_exe(). */
+};
+
+static struct mpsl_context _context;
+
+/* Update _context.return_param with parameters for the next timeslot. */
+static void schedule_next_timeslot(void)
+{
+	_context.return_param.callback_action =
+		MPSL_TIMESLOT_SIGNAL_ACTION_REQUEST;
+	_context.return_param.params.request.p_next =
+		&_context.timeslot_request;
+}
+
+static mpsl_timeslot_signal_return_param_t *timeslot_callback(
+		mpsl_timeslot_session_id_t session_id,
+		uint32_t signal)
+{
+	__ASSERT_NO_MSG(session_id == _context.session_id);
+
+	switch (signal) {
+	case MPSL_TIMESLOT_SIGNAL_START:
+		if (_context.op_desc->handler(_context.op_desc->context)
+				== FLASH_OP_DONE) {
+			_context.status = 0;
+			_context.return_param.callback_action =
+				MPSL_TIMESLOT_SIGNAL_ACTION_END;
+			k_sem_give(&_context.sem);
+		} else {
+			schedule_next_timeslot();
+		}
+
+		break;
+
+	case MPSL_TIMESLOT_SIGNAL_SESSION_IDLE:
+	case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+	case MPSL_TIMESLOT_SIGNAL_CANCELLED:
+		return NULL;
+
+	case MPSL_TIMESLOT_SIGNAL_BLOCKED:
+		schedule_next_timeslot();
+		break;
+
+	default:
+		__ASSERT(false, "unexpected signal: %u", signal);
+		return NULL;
+	}
+
+	return &_context.return_param;
+}
+
+int nrf_flash_sync_init(void)
+{
+	LOG_DBG("");
+	return k_sem_init(&_context.sem, 0, 1);
+}
+
+void nrf_flash_sync_set_context(uint32_t duration)
+{
+	LOG_DBG("duration: %u", duration);
+	_context.duration = duration;
+}
+
+bool nrf_flash_sync_is_required(void)
+{
+	return mpsl_is_initialized();
+}
+
+int nrf_flash_sync_exe(struct flash_op_desc *op_desc)
+{
+	LOG_DBG("");
+
+	uint32_t ret = mpsl_timeslot_session_open(timeslot_callback,
+			&_context.session_id);
+	if (ret < 0) {
+		LOG_ERR("mpsl_timeslot_session_open failed: %d", ret);
+		return -ENOMEM;
+	}
+
+	mpsl_timeslot_request_t *req = &_context.timeslot_request;
+	req->request_type = MPSL_TIMESLOT_REQ_TYPE_EARLIEST;
+	req->params.earliest.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE;
+	req->params.earliest.priority = MPSL_TIMESLOT_PRIORITY_NORMAL;
+	req->params.earliest.length_us = _context.duration;
```

Comment:
From what I can see, `_context.duration` cannot be trusted for write operations. We need to do the calculations ourselves.

For writes, the upper layer sets the value 7500 which may or may not be sufficient.

Similar for erase, the timeout is set to the time required to erase all pages. This gives bad scheduling performance. 
For partial erasing, the time required is also multiplied by 2 as it takes context switching into account.

Reply:
There is no way to do any calculations on this level. There is absolutely no information about the kind of operation that is preformed.

Reply:
I see. I'm not sure how to come around this limitation. @nvlsianpu  do you have some input on how this can be solved without requiring to wait for the next upmerge?

Reply:
I think `duration` is given for one page. The timeout for the complete set of operations is `FLASH_TIMEOUT_MS`.